src/SourceBuild/content/test/Microsoft.DotNet.SourceBuild.SmokeTests/ExclusionsHelper.cs

```diff
@@ -0,0 +1,177 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.Extensions.FileSystemGlobbing;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.DotNet.SourceBuild.SmokeTests;
+
+internal class ExclusionsHelper
+{
+    private const string NullSuffix = "NULL_SUFFIX";
+
+    private string _exclusionsFileName = string.Empty;
+
+    // Use this to narrow down the scope of exclusions to a specific category.
+    // For instance, setting this to "test-templates" will consider 
+    // "src/test-templates/exclusions.txt" but not "src/arcade/exclusions.txt".
+    private Regex? _exclusionRegex;
+
+    private Dictionary<string, HashSet<string>> _suffixToExclusions;
+
+    private Dictionary<string, HashSet<string>> _suffixToUnusedExclusions;
+
+    public ExclusionsHelper(string exclusionsFileName, string? exclusionRegexString = "")
+    {
+        if (exclusionsFileName is null)
+        {
+            throw new ArgumentNullException(nameof(exclusionsFileName));
+        }
+
+        _exclusionsFileName = exclusionsFileName;
+        _exclusionRegex = string.IsNullOrWhiteSpace(exclusionRegexString) ? null : new Regex(exclusionRegexString);
+        _suffixToExclusions = ParseExclusionsFile(_exclusionsFileName);
+        _suffixToUnusedExclusions = new Dictionary<string, HashSet<string>>(
+            _suffixToExclusions.ToDictionary(pair => pair.Key, pair => new HashSet<string>(pair.Value)));
+    }
+
+    internal bool IsFileExcluded(string filePath, string suffix = NullSuffix)
+    {
+        if (suffix is null)
+        {
+            throw new ArgumentNullException(nameof(suffix));
+        }
+
+        // If a specific suffix is provided, check that first. If it is not found, check the default suffix.
+        return CheckAndRemoveIfExcluded(filePath, suffix) ||
+            (suffix != NullSuffix && CheckAndRemoveIfExcluded(filePath, NullSuffix));
+    }
+
+    internal void GenerateNewBaselineFile()
+    {
+        string exclusionsFilePath = Path.Combine(BaselineHelper.GetAssetsDirectory(), _exclusionsFileName);
+
+        string[] lines = File.ReadAllLines(exclusionsFilePath);
+
+        var newLines = lines
+            .Select(line => UpdateExclusionsLine(line))
+            .Where(line => line is not null)
+            .ToArray() ?? Array.Empty<string>();
```

Comment:
Is there a need for `?? Array.Empty<string>()`. That would imply that `ToArray()` could return `null` here, which it wouldn't.